### ISSUE TITLE
How to read message with special characters (#@$!)

### ISSUE THREAD

**Issue by tendaitakas** [python-smpplib/python-smpplib#103]
Thank you for this library. I am trying to read message with:

```
def received_handler(pdu):
    print ("delivered: {} {} {}".format( pdu.receipted_message_id, pdu.short_message, pdu.source_addr)) 
```
Receiving a message like "Short message" is working fine but when I receive a french word or an sms with special characters (#@$!) the message is coming with this encoding: 

\x00S\x00o\x00m\x00e\x00 \x00w\x00e\x00i\x00r\x00d\x00 \x00c\x00a\x00r\x00a\x00c\x00t\x00e\x00r\x00s\x00 \x00%\x00#\x00 \x00*\x00 \x00\xe9\x00 \x00\xa3\x00 \x00\xe2\x00\xe6\x00 \x00\xeb\x01\x19\x00

I have tried using `pdu.short_message.decode()` but its not working.

**Comment by tendaitakas** [comment]
I needed to use `pdu.short_message.decode('latin-1')`

**Comment by eigenein** [comment]
You need to use `pdu.data_coding`, it's not always `latin-1`.

**Comment by fpalamour** [comment]
Hello,

Thank you for this lib, that's great work. I believe I'm having the same issue as @tendaitakas here. Getting pdu.short_message works great as long as there are no accents in it. If I'm correct, the short_message attribute is encoded in gsm7 and still needs to be decoded to be read correctly by my db which expects utf8. For instance, the char 'é' which is encoded 0x05 in the gsm table becomes an ENQ control character when read by my db. Do you confirm I need a gsm_decode function in order to correctly display all characters from the gsm table and if so, how would you suggest I proceed ? I have read many threads on SO but couldn't quite find a suitable solution that's short and clean.

Cheers.

**Comment by eigenein** [comment]
Hi @fpalamour,

You're correct that `short_message` is GSM7-encoded, so basically what you need is to reverse [`gsm_encode`](https://github.com/python-smpplib/python-smpplib/blob/master/smpplib/gsm.py#L54-L62). But ensure you're checking `pdu.data_coding` in advance, because it may also contain `iso-8859-1` or `utf-16-be`.

My only concern, I'm afraid we don't have in the lib yet. Correct @podshumok? I guess we need a utility function which decodes a message, taking `data_coding` into. account.

**Comment by fpalamour** [comment]
Couldn't find it in the lib myself, and, as I said, couldn't find one that pleases me either. I was looking for something short and clean but maybe I'm too picky on this. If I manage to find a suitable solution on my own to decode gsm7, would you be interested in a PR ?

**Comment by podshumok** [comment]
Yes, most of the functions in `gsm.py` deserve reversed versions: `glue_parts`, `decode`, we don't have those implemented.

**Comment by eigenein** [comment]
> If I manage to find a suitable solution on my own to decode gsm7, would you be interested in a PR ?

Sure @fpalamour

**Comment by fpalamour** [comment]
So, I have written the gsm_decode function but that's a first try and it has rooms for improvements. It seems to be working though. I'm not use to write python code so I'm sure it's possible to make it more pythonic. I'm tossing it here in case you're willing to take it from here, but I'm really not confident enough to open a PR for this.

```python
def gsm_decode(gsmtext):
    """Performs default GSM 7-bit decoding. Beware it's vendor-specific and not recommended for use."""
    try:
        ext = False
        decoded_msg = []
        for char in bytes(gsmtext, 'ascii'):
            if char == 27:
                ext = True
                continue
            decoded = char
            if ext == True:
                decoded = char + 0x80
                ext = False
            decoded_msg.append(GSM_CHARACTER_TABLE[decoded])
        return ''.join(decoded_msg)
    except Exception as e:
        raise
```

**Comment by eigenein** [comment]
It's somewhat difficult to comment on a code outside a proper PR. I'll re-open the issue for it to be visible.

Does the code also work fine with multipart messages, @fpalamour?

**Comment by fpalamour** [comment]
I believe it does yes, haven't had any issue with multipart so far.
